Metrics integration

CuriousVini opened the pull request.

Commit:
CDAP-13517-Metrics integration

chtyim commented on cdap-app-fabric/pom.xml:
```diff
@@ -205,7 +205,6 @@
       <groupId>org.apache.avro</groupId>
       <artifactId>avro</artifactId>
       <version>1.7.7</version>
```
this is a different version than the one defined in the top pom.xml file. Do you actually need 1.7.7?

CuriousVini replied:
This version was added as part of https://github.com/caskdata/cdap/commit/80b6497680458e82d98e459768b4ca5ce6251320. Some of the avro classes used in `DynamicPartitionerWithAvroTest` requires avro 1.7.7 version

chtyim replied:
It was added for testing. But now you are removing the testing, hence the question

chtyim replied:
What you can do is to actually move the test into the `cdap-app-fabric-tests` module, and have a different avro version in test scope there and then you can remove the version in the `cdap-app-fabric` module, letting it to inherit it from the parent pom.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -59,4 +67,29 @@ public boolean equals(Object o) {
     return Objects.hash(messageId, limit);
   }
+
+  public GenericRecord toGenericRecord() {
```
Missing javadoc for public method

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -32,6 +35,11 @@ public MonitorConsumeRequest(@Nullable String messageId, int limit) {
     this.limit = limit;
   }
 
+  public MonitorConsumeRequest(GenericRecord record) {
```
Missing javadoc.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -59,4 +67,29 @@ public boolean equals(Object o) {
+   * Util method to provide messageId from generic record
+   */
+  @Nullable
+  public static String getMessageId(GenericRecord record) {
```
Consider using `Optional<String>` instead of `null`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -59,4 +67,29 @@ public boolean equals(Object o) {
+    if (record.get("messageId") == null) {
+      return null;
+    }
+    return String.valueOf(record.get("messageId"));
```
Should it be `record.get("messageId").toString()` instead?

CuriousVini replied:
This method is not needed anymore.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -59,4 +67,29 @@ public boolean equals(Object o) {
+  /**
+   * Util method to provide limit from generic record
+   */
+  public static int getLimit(GenericRecord record) {
```
Why need this util method?

CuriousVini replied:
This method is not needed anymore.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -129,32 +120,113 @@ public void metadata(FullHttpRequest request, HttpResponder responder) throws Ex
     }
   }
+
+  private GenericRecord getGenericRecord(Message rawMessage) {
```
Call it `createGenericRecord`.

chtyim commented:
Do you have a JIRA for this? The commit message and the PR title should contain it.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorConsumeRequest.java:
```diff
@@ -32,6 +35,15 @@ public MonitorConsumeRequest(@Nullable String messageId, int limit) {
+   * @param record Generic record from which fields needs to be extracted
+   */
+  public MonitorConsumeRequest(GenericRecord record) {
+    this.messageId = String.valueOf(record.get("messageId"));
```
Use `record.get("messageId").toString()` instead.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorMessage.java:
```diff
@@ -16,23 +16,28 @@
-    this.messageId = messageId;
-    this.message = message;
+  public MonitorMessage (GenericRecord record) {
+    this.messageId = String.valueOf(record.get("messageId"));
```
Same as above.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/MonitorSchemas.java:
```diff
@@ -0,0 +1,83 @@
+ */
+public final class MonitorSchemas {
+
+    private MonitorSchemas() {
```
Wrong indentation. Should be 2 spaces.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -129,32 +119,112 @@ public void metadata(FullHttpRequest request, HttpResponder responder) throws Ex
+  /**
+   * Decode consume request from avro binary format
+   */
+  private Map<Utf8, GenericRecord> decodeConsumeRequest(FullHttpRequest request) throws IOException {
```
If you don't want to use `Utf8` (which is "heavier" in this particular usage context), you can add a property to the schema by using the `GenericData.setStringType(Schema, StringType)` method, telling avro to use `String` instead of `Utf8`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -129,32 +119,112 @@ public void metadata(FullHttpRequest request, HttpResponder responder) throws Ex
+      encoder.writeArrayStart();
+
+      if (!requestKeyToLocalTopic.containsKey(topicConfig)) {
+        requestKeyToLocalTopic.put(topicConfig, getTopic(topicConfig));
```
Use the `Map.computeIfAbsent` so simplify the logic:

```java
String topic = requestKeyToLocalTopic.computeIfAbsent(topicConfig, this::getTopic);
```

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -129,32 +119,112 @@ public void metadata(FullHttpRequest request, HttpResponder responder) throws Ex
+      try {
+        int limit = (int) entry.getValue().get("limit");
+        Optional<String> messageId = Optional.ofNullable(String.valueOf(entry.getValue().get("messageId")))
+          .filter(s -> !s.equals("null"));
```
Shouldn't match the `null` string. If messageId is optional in the request, make the schema field optional (by using Union type).

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -129,32 +119,112 @@ public void metadata(FullHttpRequest request, HttpResponder responder) throws Ex
+
+        fetchAndWriteMessages(encoder, buffer, chunkResponder, topic, limit, messageId.orElse(null));
+      } catch (Exception e) {
+        OUTAGE_LOG.error("Exception while sending messages for topic: {}", topic, e);
```
What types of exception would it be? Why don't want to always log the error?

CuriousVini replied:
The exceptions can be `TopicNotFoundException` or `IOException`. RuntimeMonitor will query frequently. So if there is outage, we would want reduced log frequency

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,42 +104,13 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
-      jsonWriter.endObject();
-    }
+    // Encode message and send messages
+    encodeAndSendResponse(consumeRequests, chunkResponder, buffer);
```
It's better to use a `BodyProducer` then a chunk responder. The reason is that for chunk responder, it doesn't know about the client side reading speed, hence can potentially writing too fast, causing exception raised.

chtyim replied:
You can refer to `FetchHandler` in TMS in how to a streaming one.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitorClient.java:
```diff
@@ -91,6 +104,61 @@ public RuntimeMonitorClient(String hostname, int port, HttpRequestConfig request
+
+    Map<String, Deque<MonitorMessage>> decodedMessages = new HashMap<>();
+    long entries = decoder.readMapStart();
+    while (entries > 0) {
```
since you are manually decoding map / array in avro, you should keep reading for the next number of entries once the current block is done reading. Please refer to the javadoc of the `Decoder.readMapStart()` and `Decoder.readArrayStart()` for the correct implementation.

chtyim commented on cdap-app-fabric/src/main/resources/schema/v1/MonitorConsumeRequest.avsc:
```diff
@@ -0,0 +1,11 @@
+{
+	"type" : "map",
```
Use 2 spaces instead of TAB.

CuriousVini commented:
@chtyim Thanks for the review, I have addressed the comments. Please take another pass at your convenience. Also I will add jira number to commit before squashing

CuriousVini commented:
Build is green: https://builds.cask.co/browse/CDAP-DUT6478/

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -59,21 +67,19 @@
 
   public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runnable shutdownRunnable) {
     this.cConf = cConf;
+    messageChunkSize = cConf.getInt(Constants.RuntimeMonitor.SERVER_CONSUME_CHUNK_SIZE);
```
`this.`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,76 +91,178 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
+    Decoder decoder = DecoderFactory.get().directBinaryDecoder(new ByteBufInputStream(request.content()), null);
+    Schema requestSchema = MonitorSchemas.V1.MonitorConsumeRequest.SCHEMA;
+    // Avro converts java String to UTF8 while deserializing strings. Set String type to Java String
+    GenericData.setStringType(requestSchema, GenericData.StringType.String);
```
Don't set it on every request. Just set it once in the `MonitorSchemas` class.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,76 +91,178 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
+   * Fetches all the messages for each topic provided in consumeRequests and returns iterator of topicConfig
+   * to message iterator
+   */
+  private Iterator<Map.Entry<String, CloseableIterator<Message>>> fetchAllMessages(Map<String, GenericRecord>
```
Better not to open all iterators for all topics. Why not open it one by one since you are responding one by one anyway?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
   public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
-    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+    Map<String, GenericRecord> consumeRequests = decodeConsumeRequest(request);
+    MessagesBodyProducer messagesBodyProducer = new MessagesBodyProducer(cConf, consumeRequests, messageFetcher);
```
It's probably get overlook in previous review, but the `messageFetcher` should be get from the `MessagingContext` every time when it is needed, instead of sharing the same instance across different thread.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
+      // Try to buffer up to buffer size
+      int size = 0;
+
+      while (requestsIterator.hasNext() && size < messageChunkSize) {
```
What happen if the `requestsIterator` has ended, but the `iterator` is not?

CuriousVini replied:
Good catch!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
+  private static class MessagesBodyProducer extends BodyProducer {
+    private final Iterator<Map.Entry<String, GenericRecord>> requestsIterator;
+    // caches request key to topic
+    private final Map<String, String> requestKeyToLocalTopic;
```
This map seems unnecessary since for each key from the request, you only have to resolve the local topic once and never get reused, right?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
-        }
+    @Override
+    public void handleError(@Nullable Throwable cause) {
+      while (iterator.hasNext()) {
```
Why it is a `while`?? I thought you can always just close the current `iterator`??

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitorClient.java:
```diff
@@ -91,6 +107,81 @@ public RuntimeMonitorClient(String hostname, int port, HttpRequestConfig request
+    try {
+      long entries = decoder.readMapStart();
+      while (entries > 0) {
+        String topicConfig = decoder.readString();
```
Would be easier if you use a `for` loop inside the `while`. E.g.

```java
long entries = decoder.readMapStart();
while (entries > 0) {
  for (long i = 0; i < entries; i++) {
    // Process each entry in the map. Also good to extract this part to a separate private method.
  }
  entries = decoder.readMapStart();
}
```

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java:
```diff
@@ -59,24 +67,17 @@
-  private final Map<String, String> requestKeyToLocalTopic;
 
-  public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runnable shutdownRunnable) {
+  public RuntimeHandler(CConfiguration cConf, MultiThreadMessagingContext messagingContext, Runnable shutdownRunnable) {
```
Should take the interface `MessagingContext` instead.

chtyim commented on cdap-docs/admin-manual/build.sh:
```diff
@@ -20,7 +20,7 @@ source ../vars
 
 DEFAULT_XML="../../cdap-common/src/main/resources/cdap-default.xml"
-DEFAULT_XML_MD5_HASH="dfc51846a2f0cf72962ce89de7b458b7"
+DEFAULT_XML_MD5_HASH="e49c16a4b85482ba14e19fc99cd03798"
```
You might want to rebase on top of develop and make sure the hash is still correct.

CuriousVini replied:
I just verified it. Its still the same

chtyim reviewed (commented):
Just one comment about the doc hash, the rest LGTM.

CuriousVini commented:
Squashed and Added commit Id to the commit.

CuriousVini commented:
Latest build: https://builds.cask.co/browse/CDAP-DUT6478-10

CuriousVini commented:
Build is green, merging.
